feature(Unificate Employee Patient)

lomamech opened the pull request:
- Update structure of employee table, remove all common between patient
  and employee table on Employee Table
- Update employee API on server side
- Implement the possibility to Update Employee Data
- Fix Update data on debtor and Creditor table On update Patient Data
- Update directive bhUnique add property origin to not to trigger the
  function in case you update the property hospital number

- Update integration and end to end test

closes #2282

lomamech committed:
feature(Unificate Employee Patient)
- Update structure of employee table, remove all common between patient
  and employee table on Employee Table
- Update employee API on server side
- Implement the possibility to Update Employee Data
- Fix Update data on debtor and Creditor table On update Patient Data
- Update directive bhUnique add property origin to not to trigger the
  function in case you update the property hospital number

- Update integration and end to end test

jniles commented on client/src/i18n/fr/tree.json:
```diff
@@ -17,11 +17,11 @@
     "COST_CENTER_MGMT":"Centres de coûts",
     "CREDIT_NOTE":"Note de Crédit",
-    "CREDITOR_GROUP":"Gestion de groupes crediteurs",
+    "CREDITOR_GROUP":"Gestion des groupes crediteurs",
```
:+1:

jniles commented on client/src/js/directives/bhUnique.js:
```diff
@@ -49,7 +50,15 @@ function UniqueDirective($q, UniqueValidator) {
         UniqueValidator.check(validationUrl, viewValue)
           .then(function (valueExists) {
+            // Check if hospital_no change
+            var originUpdated = viewValue !== attrs.origin;
```
This is pretty cool!

I think you might be able to do this without adding the `origin` part.  Could you try putting this at the top of your scope and seeing if it will do the same thing?
```js
var originValue = ctrl.$viewValue;

// now you can just compare to originValue
```
This will make it easier on programmers since they will not be obligated to put `origin` in the html.

lomamech replied:
There is a problem, when initializing this variable, ctrl.$viewValue is empty

jniles replied:
Hmm... I wondered if that might happen... Okay, that is fine then.  We can leave it as it is for now. Thanks for trying!

jniles commented on client/src/modules/employees/registration/employees.html:
```diff
@@ -3,7 +3,16 @@
+      <li ng-if="EmployeeCtrl.isUpdating" class="title">{{ EmployeeCtrl.employee.name }}</li>
     </ol>
+
+    <!-- ng-show to stop flickering on load -->
```
I don't think `ng-show` is used here.  Maybe you mean `ng-if`?

jniles commented on client/src/modules/employees/registration/employees.html:
```diff
@@ -3,7 +3,16 @@
+        ng-if="EmployeeCtrl.isUpdating" 
+        style="margin-left : 5px"
+        class="label label-warning">
+      <span style="text-transform : uppercase" translate>FORM.LABELS.EDIT</span>
```
There's a class for this: `text-uppercase`.

jniles commented on client/src/modules/employees/registration/employees.js:
```diff
@@ -63,15 +118,27 @@ function EmployeeController(Employees, Services, Grades, Functions, CreditorGrou
         employeeForm.$setUntouched();
         vm.employee = {};
+
+        if (!referenceId) {
```
This is cool!

jniles commented on client/src/modules/employees/templates/action.cell.html:
```diff
@@ -7,24 +7,14 @@
     <li class="dropdown-header">{{row.entity.code}}</li>
     <li>
-      <a data-method="edit" href>
+      <a data-method="edit" ui-sref="employeeEdit({id : row.entity.id})" href>
```
:+1:

jniles commented on server/controllers/medical/patients/index.js:
```diff
@@ -246,6 +247,59 @@ function lookupPatient(patientUuid) {
+        .addQuery(updateCreditor, [creditorText, creditorUuid]);
+
+      transaction.execute()
+      .then((res) => {
```
No need for this `.then` statement.  You should probably just do `return transaction.execute()`.

jniles commented on server/controllers/payroll/employees/index.js:
```diff
@@ -221,18 +228,21 @@ function update(req, res, next) {
 
   const updateCreditor = `UPDATE creditor SET ? WHERE creditor.uuid = ?`;
   const updateDebtor = `UPDATE debtor SET ? WHERE debtor.uuid = ?`;
+  const updatePatient = `UPDATE patient SET ? WHERE uuid = ?`;
```
Nice!

jniles commented on server/controllers/payroll/employees/index.js:
```diff
@@ -408,20 +427,21 @@ function find(options) {
 
   // @TODO Support ordering query
-  filters.setOrder('ORDER BY employee.display_name DESC');
+  filters.setOrder('ORDER BY patient.display_name DESC');
```
👍

jniles commented on server/models/schema.sql:
```diff
@@ -618,19 +618,13 @@ DROP TABLE IF EXISTS `employee`;
 CREATE TABLE `employee` (
   `id`            INT(11) UNSIGNED NOT NULL AUTO_INCREMENT,
   `code`          VARCHAR(20) NOT NULL,
-  `display_name`  TEXT NOT NULL,
```
:+1:

jniles reviewed (commented):
Looks pretty good to me.  Great work @lomamech.

I've made a few comments.  The logic looks good though.  Could you write an end to end test for editing the employee page?

lomamech committed:
perf(Unificate Employee Patient)
- Fix comment
- Use class property text-uppercase instead style text-transform uppercase
- return transaction.execute in function updatePatientDebCred in API for
  patient

lomamech committed:
test(Update Employee)
- Implement the test Update Employee

lomamech committed:
test(Complexe Voucher) Update the debtor select Name

jniles reviewed (approved):
Looks good to me!

bors r+
